docs/integrations/httpx.md

```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
     asyncio.run(main())
     ```
 
+## Hooks
+
+The `logfire.instrument_httpx()` also accepts the following hooks:
+
+- `request_hook`: called before sending a request.
+- `response_hook`: called after receiving a response.
+- `async_request_hook`: called before sending an async request.
+- `async_response_hook`: called after receiving an async response.
+
+You can use these hooks to add custom attributes that are not automatically captured by **Logfire**.
+
+### (Async) Request Hook
+
+The `request_hook` and `async_request_hook` hooks are called before sending a request. They receive two arguments:
+
+```py title="main.py" hl_lines="24-27"
+import asyncio
+from collections import defaultdict
+
+import httpx
+from opentelemetry.instrumentation.httpx import RequestInfo
+from opentelemetry.trace import Span
+
+import logfire
+
+
+def set_request_headers(span: Span, request: RequestInfo):
+    method, url, headers, stream, extensions = request
+    attributes: defaultdict[str, str] = defaultdict(list)
+    for key, value in headers.items():
+        key = key.lower()
+        attributes[f'http.request.header.{key}'].append(value)
+    span.set_attributes(attributes)
+
+
+def request_hook(span: Span, request: RequestInfo):
+    """This is called when using a `httpx.Client`."""
+    set_request_headers(span, request)
+
+
+async def async_request_hook(span: Span, request: RequestInfo):
+    """This is called when using a `httpx.AsyncClient`."""
```

Comment:
Two additional notes:

1. `async_request_hook` has to be a coroutine function, i.e. defined with `async def`, even if it only does sync things.
2. When instrumenting a client, the parameter is always called `request_hook`. To make it an async hook, it has to be defined as async. The type of client isn't taken into account.

I think we can automate dealing with this in the SDK. Maybe it should be an OTEL issue.